src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,141 @@
+            head = newNode;
+            tail = newNode;
+        } else {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> currentNode = findByIndex(index);
+
+        Node<T> newNode = new Node<T>(currentNode.prev, value, currentNode);
+
+        if (currentNode.prev == null) {
+            head = newNode;
+        } else {
+            currentNode.prev.next = newNode;
+        }
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return (T) findByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldValue = (T) findByIndex(index).value;
+        findByIndex(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = findByIndex(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+
+        while (currentNode != null) {
+            if ((currentNode.value == object) || (currentNode.value != null
+                    && (currentNode.value).equals(object))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node findByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode;
+        if (index < size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node node) {
+        if (node.prev == null) {
+            head = node.next;
+        } else if (node.next == null) {
+            tail = node.prev;
+        } else if (head == node) {
+            head = node.next;
+        } else if (tail == node) {
+            tail = node.prev;
```

Comment:
you have the same bodies
also node.prev == null would mean that the noode is head
```suggestion
        if (head == node) {
            head = node.next;
        } else if (tail == node) {
            tail = node.prev;
```